middleware/block_unsupported_roles.rb

```diff
@@ -0,0 +1,59 @@
+require 'mixins/client_ip'
+
+module CloudFoundry
+  module Middleware
+    class BlockUnsupportedRoles
+      def initialize(app, logger:)
+        @app                   = app
+        @logger                = logger
+      end
+
+      def call(env)
+        current_user = VCAP::CloudController::SecurityContext.current_user
+        if current_user.try(:id) && !globally_authenticated? && config_enabled? && space_supporter_and_only_space_supporter?(current_user)
+          [
+            403,
+            { 'Content-Type' => 'text/html' },
+            'You are not authorized to perform the requested action. See section \'Space Supporter Role in V2\' https://docs.cloudfoundry.org/concepts/roles.html',
+          ]
+        else
+          @app.call(env)
+        end
+      end
+
+      private
+
+      def allowed_path?(path)
+        ['/v2/info', '/'].include?(path)
+      end
+
+      def globally_authenticated?
+        roles.admin? || roles.admin_read_only? || roles.global_auditor?
+      end
+
+      def config_enabled?
+        VCAP::CloudController::Config.config.get(:temporary_enable_space_supporter_role)
+      end
+
+      def space_supporter_and_only_space_supporter?(current_user)
+        user_roles = VCAP::CloudController::SpaceManager.select(Sequel.as(VCAP::CloudController::RoleTypes::SPACE_MANAGER, :type)).where(user_id: current_user.id).
+                     union(VCAP::CloudController::SpaceDeveloper.select(Sequel.as(VCAP::CloudController::RoleTypes::SPACE_DEVELOPER, :type)).where(user_id: current_user.id)).
+                     union(VCAP::CloudController::SpaceAuditor.select(Sequel.as(VCAP::CloudController::RoleTypes::SPACE_AUDITOR, :type)).where(user_id: current_user.id)).
+                     union(VCAP::CloudController::SpaceSupporter.select(Sequel.as(VCAP::CloudController::RoleTypes::SPACE_SUPPORTER, :type)).where(user_id: current_user.id)).
+                     union(VCAP::CloudController::OrganizationManager.select(Sequel.as(VCAP::CloudController::RoleTypes::ORGANIZATION_MANAGER,
+:type)).where(user_id: current_user.id)).
+                     union(VCAP::CloudController::OrganizationBillingManager.select(Sequel.as(VCAP::CloudController::RoleTypes::ORGANIZATION_BILLING_MANAGER,
+:type)).where(user_id: current_user.id)).
+                     union(VCAP::CloudController::OrganizationAuditor.select(Sequel.as(VCAP::CloudController::RoleTypes::ORGANIZATION_AUDITOR,
+:type)).where(user_id: current_user.id)).
+                     all
+
+        user_roles.count == 1 && user_roles[0][:type] == 'space_supporter'
```

Comment:
I think this count check would allow a SpaceSupporter who is a SpaceSupporter for two or more spaces to access these endpoints.  I think it should be more like checking that all roles for the user have `:type` == `space_supporter`.  Let me know if I'm wrong in how the modeling works.

Reply:
The union operation does a unique on the set (`union all` apparently would not unique the set).

We wrote a test for this to prove to ourselves it was working. We also added a LIMIT 1 clause to all the selects as it makes it obvious and could help the query optimizer out maybe?

Reply:
Ah, got it.  Thanks.